### ISSUE TITLE
`dyn Future` suppresses unconditional recursion 

### ISSUE THREAD

**Issue by NishantJoshi00** [rust-lang/rust#111177]
There was a `unconditional_recursion` in code, and there was no warning related to it. (in case of `dyn Future`)

I tried this code:
## Code

```rust
use std::{future::Future, pin::Pin}

fn do_something() -> Pin<Box<dyn Future<Output = ()>>> {
  Box::pin(async { do_something().await })
}

```

The code mentioned above does not trigger any warning or errors. The behaviour was first caught when working with `async_trait`, where it also does a similar transformation. 

This is unique to `dyn Future` because, I tried reproducing this with other possible `Future` recursions but, they were instantly caught.

## Examples

With Async Function
```rust
async fn do_something() {
  do_something().await
}

/*

error[E0733]: recursion in an `async fn` requires boxing
 --> src/main.rs:8:25
  |
8 | async fn do_something() {
  |                         ^ recursive `async fn`
  |
  = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`
  = note: consider using the `async_recursion` crate: https://crates.io/crates/async_recursion

*/
```

With `impl Future`
```rust

fn do_something() -> Pin<Box<impl Future<Output = ()>>> {
  Box::pin( async { do_something().await } )
}

/*

error[E0720]: cannot resolve opaque type
  --> src/main.rs:12:31
   |
12 | fn do_something() -> Pin<Box<impl Future<Output = ()>>> {
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ recursive opaque type
13 |     Box::pin(async { do_something().await })
   |     -----------------------------------------
   |     |                |
   |     |                async closure captures itself here
   |     returning here with type `Pin<Box<[async block@src/main.rs:13:14: 13:45]>>`

*/

```

**Comment by drewtato** [comment]
This also happens with non-async functions.

```rs
fn recur() {
    Box::new(recur)();
}
```

The only message is from clippy saying the box is unnecessary.

Also notable is that an unconditionally recursive async function is not necessarily incorrect. The original one from Nishant that only calls itself is as incorrect as sync recursion[^1], but one that does some async work and then calls itself could be useful, just not idiomatic or performant.

[^1]: It basically _is_ sync recursion, in the first call to `poll`

```rs
use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use tokio::select;
use tokio::time::sleep;

fn recur() -> Pin<Box<dyn Future<Output = ()>>> {
    Box::pin(async {
        sleep(Duration::from_millis(10)).await;
        println!("Slept once");
        recur().await;
    })
}

#[tokio::main]
async fn main() {
    select! {
        () = recur() => unreachable!(),
        () = sleep(Duration::from_millis(100)) => println!("Done"),
    }
}
```

This prints a few times and exits. Cancelling futures is much more acceptable than cancelling threads, which would need to be done for a program with a sync recursive function to exit cleanly. I don't think this is a common use case, though, so I'm still in favor of the lint. It's likely that functions of this form are still mistakes.

Last thing: the current error is because plain recursion is completely impossible in async functions.

```rs
async fn recur_never() {
    if false {
        recur_never().await;
    }
}
```

I'm pretty sure this is because it would create an infinite-sized type. It's not concerned with the correctness of the recursion. It's not the async counterpart to `unconditional_recursion`. If it was possible, this would be allowed.

The other error with `impl Future` is because it creates an infinitely recursive type. This is not really related to async.

**Comment by NishantJoshi00** [comment]
I am still not sure, how `impl Future` here is creating a infinitely recursive type, as the type signature of the output, is fixed as we are awaiting it inside, in-turn have a fixed type signature. 
> ```
> /*
> 
> error[E0720]: cannot resolve opaque type
>   --> src/main.rs:12:31
>    |
> 12 | fn do_something() -> Pin<Box<impl Future<Output = ()>>> {
>    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ recursive opaque type
> 13 |     Box::pin(async { do_something().await })
>    |     -----------------------------------------
>    |     |                |
>    |     |                async closure captures itself here
>    |     returning here with type `Pin<Box<[async block@src/main.rs:13:14: 13:45]>>`
> 
> */
> ```

The way I see it, it's referencing a future that is present somewhere on the heap, again and again. This definitely can cause linearly high amount of allocation. But, in this case the compiler mentions that it's a `recursive opaque type`, can you provide a tad bit more context on how type system is preventing this from happening and the same is not the case for for `dyn Future`. I am definitely curious as to how the compiler itself is understanding it. If you have any resource that could provide insight about this situation, it would definitely be helpful

**Comment by drewtato** [comment]
> I am still not sure, how `impl Future` here is creating a infinitely recursive type

A reminder: async blocks (and functions) are compiled into state machines that store any values that are held across an await point, including the future being awaited. It's like an enum where each variant is a tuple of the values being held.

The `impl Trait` syntax is only an API concern. You're saying "it is an error to interact with the value in any way besides what `Trait` provides". But when compiled into a binary, it will produce the same code as if the real type was used instead. The type system needs to eventually figure out the real type, even though it's opaque to you. So you could imagine the return of the function becoming `Pin<Box<[async block that stores impl Future<Output = ()>]>>`. However, `impl Future<Output = ()>` also needs to be replaced, resulting in:

```rs
Pin<Box<
    [async block that stores Pin<Box<
        [async block that stores Pin<Box<
            [async block that stores /* clearly a problem */]
        >>]
    >>]
>>
```

This is a type that has a finite size but an infinite name.

The [error code](https://doc.rust-lang.org/stable/error_codes/E0720.html) actually has a pretty nice example that doesn't involve async:

```rs
fn make_recursive_type() -> impl Sized {
    [make_recursive_type(), make_recursive_type()]
}
```

This returns an array of type `[[[... ; 2]; 2]; 2]`, which is either infinite- or zero-sized. But it has an infinite name, and therefore cannot be type checked. `Sized` is being used here as essentially any trait.

`dyn Future<Output = ()>` is already a real type, so this process does not happen there. The type of the async block is erased (i.e. moved to runtime) when it's converted into `dyn Future`.

**Comment by NishantJoshi00** [comment]
@drewtato that makes sense! Returning to the original issue, this isn't just about the `dyn Future`; and as you mentioned, the type system cannot catch it as `dyn Future<Output = ()>` is already a solid type. this leaves out the possibility of it being caught by the type system. Also, the use case that you mentioned where `Box<fn() -> ()>()` calling the function after boxing also didn't throw any warning. But, if we follow the compiler with this we can eventually trace it as recursion. While, in our case, this won't be possible.

But having a lint that can provide a little bit of insight on recursive futures that are created in line would definitely be helpful.

**Comment by tyilo** [comment]
I just encountered a similar situation (also with `async-trait`) which can be reduced to:

```rust
use std::sync::Arc;
use async_trait::async_trait;

#[async_trait]
trait Trait {
    async fn f(&self);
}

#[async_trait]
impl Trait for () {
    async fn f(&self) {
        println!("foo");
    }
}

#[async_trait]
impl<T: Trait + Send + Sync> Trait for Arc<T> {
    async fn f(&self) {
        // Both of these loop forever:
        // self.f().await
        // (*self).f().await
        (**self).f().await
    }
}
```

It's quite easy to see that `self.f().await` will not work, however it is more tricky to realize that `(*self).f().await` doesn't work either.
